### PR TITLE
Add generic Bootstrap Token constants and helpers to client-go

### LINKED ISSUE #114: [proposal] Add generic Bootstrap Token constants and helpers to client-go
Features issue: https://github.com/kubernetes/features/issues/130
Proposal: https://github.com/kubernetes/community/pull/189
Initial implementation PR: https://github.com/kubernetes/kubernetes/issues/36101

Why should this be in client-go?
Well, I and @sttts had a discussion about this in #kubernetes-dev on Slack:

Slack thread:
<details>

```
luxas [14:25] 
What do you think about the bootstrap api and the util functions it needs?
@sttts Can it be in client-go?
I think that would make a lot of sense, because this is code people are gonna build on top of

sttts [14:26] 
@luxas we don't have many alternatives I fear

luxas [14:26] 
and the fear is?
just that there are no other alternatives or something else?

sttts [14:26] 
client-go is a bit (too much) our catch-all for packages where we don't know where to put them (edited)

luxas [14:27] 
yeah, I see

sttts [14:28] 
which kind of util funcs?

luxas [14:28] 
validation mostly
regexes for valid tokens
maybe generation of a random one
and for example, a function that builds a full token from an id and a secret, that's just `fmt.Sprintf("%s:%s", id, secret)`
but that way it's clear that the char in the middle isn't `.` or `-` or whatever
a function that checks whether the token has expired
and so on
but all these would in that case be super-generic
only the essentials

sttts [14:33] 
sounds like something for tools/bootstrap

luxas [14:34] 
might be
right now the constants live in `pkg/bootstrap/api`

sttts [14:35] 
which doesn't make sense if it cannot be shared

luxas [14:35] 
and since this isn't a "real" api, it shouldn't be in `client-go/pkg/api{,s}`

sttts [14:35] 
I think client-go/tools/bootstrap is good

luxas [14:35] 
but `client-go/tools/bootstrap` sgtm

sttts [14:36] 
what is left in pkg/bootstrap?
after moving that

luxas [14:36] 
nothing

sttts [14:36] 
+1

luxas [14:36] 
now only the constants are there
most of the logic is "hardcoded" in `pkg/controller/bootstrap`, but the most generic utils would in this case be moved out
and only the controller logic would be left in `pkg/controller/bootstrap`

sttts [14:38] 
makes sense

luxas [14:38] 
then the main and first consumers of `client-go/tools/bootstrap` will be `cmd/kubeadm/app`, `pkg/controller/bootstrap` and `apiserver/plugin/pkg/authenticator/token/bootstrap`
but I expect kops to use this in some way or another in the future as well
```
</details>

The proposal is to:
 - move the constants from `k8s.io/kubernetes/pkg/bootstrap/api` to `k8s.io/client-go/tools/bootstrap/(token/)api`
 - factor out the most essential validation, generation and information getters about the token to `k8s.io/client-go/tools/bootstrap/(token/)utils` or similar

Then 
 - `k8s.io/kubernetes/cmd/kubeadm/app`
 - `k8s.io/kubernetes/pkg/controller/bootstrap`
 - `k8s.io/kubernetes/apiserver/plugin/pkg/authenticator/token/bootstrap` (see: https://github.com/kubernetes/kubernetes/pull/41281)

can use those common util funcs and it's generally easier to build upon the bootstrap token functionality.

This should preferably make v1.6, since we're aiming for beta for this functionality

cc @jbeda @mikedanese @roberthbailey @pires @dmmcquay @errordeveloper @deads2k @smarterclayton @liggitt @sttts @justinsb @ericchiang

### PR DESCRIPTION
**Pull request by guangxuli**
**What this PR does / why we need it**:
per https://github.com/kubernetes/client-go/issues/114

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

fix  https://github.com/kubernetes/client-go/issues/114

**Special notes for your reviewer**:

**Release note**:

```release-note
none
```

### CODE REVIEW

File: staging/src/k8s.io/client-go/tools/bootstrap/token/api/doc.go [hunk 1]
```diff
@@ -14,7 +14,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-// Package api (pkg/bootstrap/api) contains constants and types needed for
+// Package api (bootstrap/token/api) contains constants and types needed for
 // bootstrap tokens as maintained by the BootstrapSigner and TokenCleaner
 // controllers (in pkg/controller/bootstrap)
-package api // import "k8s.io/kubernetes/pkg/bootstrap/api"
+package api // import "k8s.io/kubernetes/client-go/tools/bootstrap/token/api"
```
Comment (sttts) [inline staging/src/k8s.io/client-go/tools/bootstrap/token/api/doc.go]: k8s.io/client-go

Comment (guangxuli) [reply staging/src/k8s.io/client-go/tools/bootstrap/token/api/doc.go]: oh, yes,  thanks@sttts.

**Comment by guangxuli** [comment]
@sttts @jbeda @mattmoyer PTAL, thanks

**Review by luxas — COMMENTED** [review]
I would hold this until the v1.10 cycle; this is too late for v1.9 IMO

Very happy to see this in v1.10 though

**Comment by guangxuli** [comment]
@luxas sure, thanks. :+1:

**Comment by luxas** [comment]
@mattmoyer @jbeda PTAL

**Comment by luxas** [comment]
@guangxuli Please rebase this when you're ready and we'll re-evaluate this one

**Comment by guangxuli** [comment]
@luxas done, thanks.

File: staging/src/k8s.io/client-go/tools/bootstrap/token/api/doc.go [hunk 2]
```diff
@@ -14,7 +14,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-// Package api (pkg/bootstrap/api) contains constants and types needed for
+// Package api (bootstrap/token/api) contains constants and types needed for
```
Comment (sttts) [inline staging/src/k8s.io/client-go/tools/bootstrap/token/api/doc.go]: tools/bootstrap...

**Comment by sttts** [comment]
I still have my doubt that client-go is the right place. But we don't have a better place than this either. So I think I can live with this. Giving @jbeda @mattmoyer @caesarxuchao a chance to look at this.

**Comment by guangxuli** [comment]
@sttts sure.
kindly ping @caesarxuchao @jbeda @mattmoyer expect your suggestions.

**Comment by jbeda** [comment]
/lgtm
/approve no-issue

Also not sure the right place to put stuff like this.  If we had better ways to have lots of little packages that might be good.  `k8s.io/api` might be there other place but there are no helpers there at all.  `client-go` is the place to put API specific helpers.

One thing we could do -- move the constants to `k8s.io/api` and have the helpers in `client-go`.  But at this point it seems like diminishing returns.

**Comment by sttts** [comment]
@guangxuli tests are not happy

**Comment by errordeveloper** [comment]
@guangxuli please rebase.

**Comment by guangxuli** [comment]
@errordeveloper thanks, have rebased, just trigger another round test...

**Comment by errordeveloper** [comment]
There is a script in hack directory that updates build files, have you
tried it already?

**Comment by guangxuli** [comment]
@errordeveloper actually, yes, but there still a bazel test failure.  I have to go deep into more details.

**Comment by guangxuli** [comment]
@sttts @errordeveloper @jbeda finally, all tests happy, PTAL again.

**Pull request merged**